fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -800,57 +833,121 @@ public CompletableFuture<RecordCursorContinuation> rebalancePartitions(RecordCur
-                            .setPartitionKey(getPartitionKey(partitionInfo))
-                            .setRepartitionDocCount(count);
-
-                    long startTimeNanos = System.nanoTime();
-                    LuceneRecordCursor luceneRecordCursor = getOldestNDocuments(partitionInfo, groupingKey, count);
-
-                    return moveDocsFromPartition(partitionInfo, groupingKey, maxPartitionId, luceneRecordCursor)
-                            .thenApply(movedCount -> {
-                                state.context.record(LuceneEvents.Events.LUCENE_REBALANCE_PARTITION, System.nanoTime() - startTimeNanos);
-                                state.context.recordSize(LuceneEvents.SizeEvents.LUCENE_REBALANCE_PARTITION_DOCS, movedCount);
-                                return Pair.of(
-                                        movedCount,
-                                        Math.max(partitionInfo.getCount() - movedCount - indexPartitionHighWatermark, 0));
-                            });
+                    if (previous != null && previous.getCount() + currentPartitionCount <= indexPartitionHighWatermark) {
+                        // merge current into previous
+                        repartitioningContext.cursor = getOldestNDocuments(partitionInfo, groupingKey, repartitioningContext.countToMove);
+                    } else if (next != null && next.getCount() + currentPartitionCount <= indexPartitionHighWatermark) {
+                        // merge current into next
+                        repartitioningContext.removingOldest = false;
+                        repartitioningContext.cursor = getNewestNDocuments(partitionInfo, groupingKey, repartitioningContext.countToMove);
+                    } else if (previous != null && next != null && (2 * indexPartitionHighWatermark - previous.getCount() - next.getCount()) >= currentPartitionCount ) {
+                        // split current into next and previous
+                        // for this iteration, we do a partial move into previous; next iteration(s) would eventually
+                        // move the rest to next
+                        int previousCapacity = indexPartitionHighWatermark - previous.getCount();
+                        repartitioningContext.countToMove = Math.min(previousCapacity, repartitionDocumentCount);
+                        if (currentPartitionCount > repartitioningContext.countToMove) {
+                            repartitioningContext.countToMove++;
+                            repartitioningContext.newBoundaryRecordPresent = true;
+                        }
+                        repartitioningContext.emptyingPartition = false;
+                        repartitioningContext.cursor = getOldestNDocuments(partitionInfo, groupingKey, repartitioningContext.countToMove);
+                    } else {
+                        // partition is under low-watermark, but no capacity around it to absorb it
+                        continue;
```

Comment:
does `repartitioningCount` need to get reset if this happens? 
I think the flow would have to be something like:
1. see a partition that you should empty
2. all of the if/else-if above return false so it gets to here
3. next partition should not be emptied.

The only other field that is set is the `countToMove` which always gets set.
It does seem a bit brittle though to have it living between partitions, even if resetting `emptyingPartition` would be sufficient for right now.